File: stepik-union/src/main/java/org/stepik/core/ui/StepDescriptionUtils.java
package org.stepik.core.ui;

import com.intellij.openapi.diagnostic.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.api.objects.submissions.Choice;
import org.stepik.api.objects.submissions.Column;
import org.stepik.api.objects.submissions.Reply;
import org.stepik.core.courseFormat.StepNode;
import org.stepik.core.courseFormat.StepType;
import org.w3c.dom.Node;
import org.w3c.dom.html.HTMLInputElement;
import org.w3c.dom.html.HTMLSelectElement;
import org.w3c.dom.html.HTMLTextAreaElement;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static java.util.Collections.emptyList;

class StepDescriptionUtils {
    private static final Logger logger = Logger.getInstance(StepDescriptionUtils.class);

    @Nullable
    static Reply getReply(
            @NotNull StepNode stepNode, @NotNull StepType type,
            @NotNull Elements elements,
            @Nullable String data) {
        Reply reply = new Reply();
        switch (type) {
            case CHOICE:
                List<Boolean> choices = getChoiceData(elements);
                reply.setChoices(choices);
                break;
            case STRING:
                String text = getStringData(elements);
                reply.setText(text);
                break;
            case FREE_ANSWER:
                text = getStringData(elements);
                reply.setText(text);
                reply.setAttachments(emptyList());
                break;
            case NUMBER:
                String number = getStringData(elements);
                reply.setNumber(number);
                break;
            case DATASET:
                String dataset;
                if (data == null) {
                    dataset = getStringData(elements);
                } else {
                    dataset = data;
                }
                reply.setFile(dataset);
                break;
            case TABLE:
                List tableChoices = getChoices(elements);
                reply.setChoices(tableChoices);
                break;
            case FILL_BLANKS:
                List<String> blanks = getBlanks(elements);
                reply.setBlanks(blanks);
                break;
            case SORTING:
            case MATCHING:
                List<Integer> ordering = getOrderingData(elements);
                reply.setOrdering(ordering);
                break;
            case MATH:
                String formula = getStringData(elements);
                reply.setFormula(formula);
                break;
            default:
                logger.warn("Unknown step type tried sending: " + type);
                return null;
        }
        stepNode.setLastReply(reply);
        return reply;
    }

    private static void forEachInputElement(@NotNull Elements elements, Consumer<HTMLInputElement> consumer) {
        for (Node node : elements) {
            if (node instanceof HTMLInputElement) {
                HTMLInputElement element = (HTMLInputElement) node;
                consumer.accept(element);
                element.setDisabled(true);
            }
        }
    }

    private static void disableAllInputs(@NotNull Elements elements) {
        for (Node node : elements) {
            if (node instanceof HTMLInputElement) {
                HTMLInputElement element = (HTMLInputElement) node;
                element.setDisabled(true);
            } else if (node instanceof HTMLTextAreaElement) {
                HTMLTextAreaElement element = (HTMLTextAreaElement) node;
                element.setDisabled(true);
            }
        }
    }

    @NotNull
    private static List<Boolean> getChoiceData(@NotNull Elements elements) {
        List<Boolean> choices = new ArrayList<>();

        forEachInputElement(elements, element -> {
            if ("option".equals(element.getName())) {
                choices.add(element.getChecked());
            }
        });

        return choices;
    }

    @NotNull
    private static List<Integer> getOrderingData(@NotNull Elements elements) {
        List<Integer> ordering = new ArrayList<>();

        forEachInputElement(elements, element -> {
            if ("index".equals(element.getName())) {
                String indexAttr = element.getValue();
                ordering.add(Integer.valueOf(indexAttr));
            }
        });
        return ordering;
    }

    private static String getStringData(@NotNull Elements elements) {
        disableAllInputs(elements);
        return elements.getInputValue("text");
    }

    @NotNull
    private static List<Choice> getChoices(@NotNull Elements elements) {
        Map<String, List<Column>> choices = new HashMap<>();
        List<String> rows = new ArrayList<>();

        forEachInputElement(elements, element -> {
            String type = element.getType();

            if ("checkbox".equals(type) || "radio".equals(type)) {
                List<Column> columns = choices.computeIfAbsent(element.getName(), k -> {
                    rows.add(k);
                    return new ArrayList<>();
                });
                Column column = new Column(element.getValue(), element.getChecked());
                columns.add(column);
            }
        });

        return choices.entrySet().stream()
                .map(entry -> new Choice(entry.getKey(), entry.getValue()))
                .sorted(Comparator.comparingInt(choice -> rows.indexOf(choice.getNameRow())))
                .collect(Collectors.toList());
    }

    private static List<String> getBlanks(@NotNull Elements elements) {
        List<String> blanks = new ArrayList<>();

        for (Node node : elements) {
            if (node instanceof HTMLInputElement) {
                HTMLInputElement element = (HTMLInputElement) node;
                String type = element.getType();
                if ("text".equals(type)) {
                    blanks.add(element.getValue());
                }
                element.setDisabled(true);
            } else if (node instanceof HTMLSelectElement) {
                HTMLSelectElement element = (HTMLSelectElement) node;
                blanks.add(element.getValue());
                element.setDisabled(true);
            }
        }

        return blanks;
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/ui/StepDescriptionUtils.kt
package org.stepik.core.ui

import org.stepik.api.objects.submissions.Attachment
import org.stepik.api.objects.submissions.Choice
import org.stepik.api.objects.submissions.Column
import org.stepik.api.objects.submissions.Reply
import org.stepik.core.courseFormat.StepNode
import org.stepik.core.courseFormat.StepType
import org.w3c.dom.html.HTMLInputElement
import org.w3c.dom.html.HTMLSelectElement
import org.w3c.dom.html.HTMLTextAreaElement
import java.util.*
import java.util.Collections.emptyList
import java.util.function.Consumer

internal object StepDescriptionUtils {

    fun getReply(
            stepNode: StepNode,
            type: StepType,
            elements: Elements,
            data: String?): Reply? {
        val reply = Reply()
        when (type) {
            StepType.CHOICE -> {
                val choices = getChoiceData(elements)
                reply.setChoices(choices)
            }
            StepType.STRING -> {
                val text = getStringData(elements)
                reply.setText(text)
            }
            StepType.FREE_ANSWER -> {
                val text = getStringData(elements)
                reply.setText(text)
                reply.setAttachments(emptyList<Attachment>())
            }
            StepType.NUMBER -> {
                val number = getStringData(elements)
                reply.setNumber(number)
            }
            StepType.DATASET -> {
                val dataset: String = data ?: getStringData(elements)
                reply.setFile(dataset)
            }
            StepType.TABLE -> {
                val tableChoices = getChoices(elements)
                reply.setChoices(tableChoices)
            }
            StepType.FILL_BLANKS -> {
                val blanks = getBlanks(elements)
                reply.setBlanks(blanks)
            }
            StepType.SORTING, StepType.MATCHING -> {
                val ordering = getOrderingData(elements)
                reply.setOrdering(ordering)
            }
            StepType.MATH -> {
                val formula = getStringData(elements)
                reply.setFormula(formula)
            }
            else -> return null
        }
        stepNode.lastReply = reply
        return reply
    }

    private fun forEachInputElement(elements: Elements, consumer: Consumer<HTMLInputElement>) {
        for (node in elements) {
            if (node is HTMLInputElement) {
                consumer.accept(node)
                node.disabled = true
            }
        }
    }

    private fun disableAllInputs(elements: Elements) {
        for (node in elements) {
            if (node is HTMLInputElement) {
                node.disabled = true
            } else if (node is HTMLTextAreaElement) {
                node.disabled = true
            }
        }
    }

    private fun getChoiceData(elements: Elements): List<Boolean> {
        val choices = ArrayList<Boolean>()

        forEachInputElement(elements, Consumer {
            if (it.name == "option") {
                choices.add(it.checked)
            }
        })

        return choices
    }

    private fun getOrderingData(elements: Elements): List<Int> {
        val ordering = ArrayList<Int>()

        forEachInputElement(elements, Consumer {
            if (it.name == "index") {
                val indexAttr = it.value
                ordering.add(Integer.valueOf(indexAttr))
            }
        })
        return ordering
    }

    private fun getStringData(elements: Elements): String {
        disableAllInputs(elements)
        return elements.getInputValue("text")
    }

    private fun getChoices(elements: Elements): List<Choice> {
        val choices = HashMap<String, MutableList<Column>>()
        val rows = ArrayList<String>()

        forEachInputElement(elements, Consumer {
            val type = it.type

            if (type in listOf("checkbox", "radio")) {
                val columns = choices.computeIfAbsent(it.name) {
                    rows.add(it)
                    mutableListOf()
                }
                val column = Column(it.value, it.checked)
                columns.add(column)
            }
        })

        return choices.entries
                .map { entry -> Choice(entry.key, entry.value) }
                .sortedBy { rows.indexOf(it.nameRow) }
    }

    private fun getBlanks(elements: Elements): List<String> {
        val blanks = ArrayList<String>()

        for (node in elements) {
            if (node is HTMLInputElement) {
                if (node.type == "text") {
                    blanks.add(node.value)
                }
                node.disabled = true
            } else if (node is HTMLSelectElement) {
                blanks.add(node.value)
                node.disabled = true
            }
        }

        return blanks
    }
}
